### PR TITLE
chore: revise ESLint built-in configuration

### PR DESCRIPTION
**Pull request by jviotti**
There are a lot of new rules since the last time I revised the ESLint
rules documentation.

I've updated the main `.eslintrc.yml` to include some newer additions,
plus I added another ESLint configuration file inside `tests`, so we can
add some stricted rules to the production code while relaxing them for
the test suite (due to the fact that Mocha is not very ES6 friendly and
Angular tests require a bit of dark magic to setup).

This is a summary of the most important changes:

- Disallow "magic numbers"

These should now be extracted to constants, which forces us to think of
a good name for them, and thus make the code more self-documenting (I
had to Google up the meaning of some existing magic numbers, so I guess
this will be great for readability purposes).

- Require consistent `return` statements

Some functions relied on JavaScript relaxed casting mechanism to work,
which now have explicit return values. This flag also helped me detect
some promises that were not being returned, and therefore risked not
being caught by the exception handlers in case of errors.

- Disallow redefining function arguments

Immutability makes functions easier to reason about.

- Enforce JavaScript string templates instead of string concatenation

We were heavily mixing boths across the codebase.

There are some extra rules that I tweaked, however most of codebase
changes in this commit are related to the rules mentioned above.

Signed-off-by: Juan Cruz Viotti <jviotti@openmailbox.org>

### CODE REVIEW

**Comment by lurch** [comment]
Looks like you've been 'relaxying' your spelling too. LOL

**Comment by jviotti** [comment]
lol, fixed!

File: lib/gui/models/store.js [hunk 1]
```diff
@@ -91,7 +105,8 @@ const storeReducer = (state, action) => {
 
       const newState = state.set('availableDrives', Immutable.fromJS(action.data));
```
Comment (lurch) [inline lib/gui/models/store.js]: I've never written any Redux code before, but should the first parameter to these `.set` calls also be constants, rather than just strings, to cut down on the possibility of any typos?

File: lib/gui/models/store.js [hunk 2]
```diff
@@ -354,7 +369,7 @@ module.exports = _.merge(redux.createStore(
       // otherwise we will be basically erasing the property
       // we aim the keep serialising the in future.
```
Comment (lurch) [inline lib/gui/models/store.js]: "aim to keep" ?

File: lib/gui/models/supported-formats.js [hunk 3]
```diff
@@ -115,7 +115,7 @@ SupportedFormats.service('SupportedFormatsModel', function() {
    * }
    */
   this.isSupportedImage = (imagePath) => {
-    const extension = path.extname(imagePath).slice(1).toLowerCase();
+    const extension = _.join(_.tail(path.extname(imagePath)), '').toLowerCase();
```
Comment (lurch) [inline lib/gui/models/supported-formats.js]: IMHO this is _much_ more unreadable. What about using `replace('.', '')` instead?

Also, `.${extension}` is used again at the bottom of the function, so I guess you could maybe do:
```
const extname = path.extname(imagePath);
const extension = extname.replace('.', '').toLowerCase();
```
and then replace `.${extension}` with `extname`. Should there also be some check for `extension` being empty?

File: lib/gui/modules/analytics.js [hunk 4]
```diff
@@ -129,11 +129,8 @@ analytics.service('AnalyticsService', function($log, $window, $mixpanel, Setting
 
     }
 
-    if (data) {
-      message += ` (${JSON.stringify(data)})`;
-    }
-
-    this.logDebug(message);
+    const debugMessage = data ? `${message} (${JSON.stringify(data)})` : message;
```
Comment (lurch) [inline lib/gui/modules/analytics.js]: IMHO the old version was more readable, but I guess this is the price you pay for banning string concatenation ;)

File: lib/gui/modules/drive-scanner.js [hunk 5]
```diff
@@ -34,16 +34,17 @@ const driveScanner = angular.module(MODULE_NAME, [
 
 driveScanner.factory('DriveScannerService', (SettingsModel) => {
   const DRIVE_SCANNER_INTERVAL_MS = 2000;
+  const FIRST_SCAN_DELAY_MS = 0;
```
Comment (lurch) [inline lib/gui/modules/drive-scanner.js]: It would be nice if this was also prefixed with `DRIVE_SCANNER_` to match the other constant here.

File: lib/gui/modules/drive-scanner.js [hunk 6]
```diff
@@ -52,7 +53,9 @@ driveScanner.factory('DriveScannerService', (SettingsModel) => {
 
         return drive;
       });
+    })
 
+    .map((drives) => {
```
Comment (lurch) [inline lib/gui/modules/drive-scanner.js]: I'm not familiar with the coding style used here (so I can't quite tell what this is doing) but how come the old code had two `map` functions, but the new code has three?

File: lib/gui/os/dialog/services/dialog.js [hunk 7]
```diff
@@ -113,20 +113,28 @@ module.exports = function($q, SupportedFormatsModel) {
    * });
    */
   this.showWarning = (options) => {
+    const BUTTONS = [
+      options.confirmationLabel,
+      options.rejectionLabel
+    ];
+
+    const BUTTON_CONFIRMATION_INDEX = _.indexOf(BUTTONS, options.confirmationLabel);
+    const BUTTON_REJECTION_INDEX = _.indexOf(BUTTONS, options.rejectionLabel);
```
Comment (lurch) [inline lib/gui/os/dialog/services/dialog.js]: This looks horrible :-(

Feels like there must be a nicer way to do this...?

File: lib/gui/os/dialog/services/dialog.js [hunk 8]
```diff
@@ -113,20 +113,28 @@ module.exports = function($q, SupportedFormatsModel) {
...
+
     return $q((resolve) => {
       electron.remote.dialog.showMessageBox(currentWindow, {
         type: 'warning',
         buttons: [
```
Comment (lurch) [inline lib/gui/os/dialog/services/dialog.js]: Given you've constructed the array above, could this just be `buttons: BUTTONS` ?

File: lib/gui/os/dialog/services/dialog.js [hunk 9]
```diff
@@ -113,20 +113,28 @@ module.exports = function($q, SupportedFormatsModel) {
...
-        defaultId: 1,
-        cancelId: 1,
+        defaultId: BUTTON_REJECTION_INDEX,
+        cancelId: BUTTON_REJECTION_INDEX,
         title: 'Attention',
         message: options.title,
         detail: options.description
       }, (response) => {
-        return resolve(response === 0);
+        return resolve(response === BUTTON_CONFIRMATION_INDEX);
```
Comment (lurch) [inline lib/gui/os/dialog/services/dialog.js]: It means changing an integer-comparison to a string-comparison, but I guess `return resolve(BUTTONS[response] === options.confirmationLabel);` would work here?

File: lib/gui/os/dialog/services/dialog.js [hunk 10]
```diff
@@ -113,20 +113,28 @@ module.exports = function($q, SupportedFormatsModel) {
    * });
    */
   this.showWarning = (options) => {
+    const BUTTONS = [
+      options.confirmationLabel,
```
Comment (lurch) [inline lib/gui/os/dialog/services/dialog.js]: Should we be setting defaults for `.confirmationLabel` and `.rejectionLabel` in case they don't exist?

File: lib/gui/os/window-progress/services/window-progress.js [hunk 11]
```diff
@@ -47,11 +44,14 @@ module.exports = function() {
    * OSWindowProgressService.set(85);
    */
   this.set = (percentage) => {
-    if (percentage > 100 || percentage < 0) {
+    const PERCENTAGE_MINIMUM = 0;
+    const PERCENTAGE_MAXIMUM = 100;
```
Comment (lurch) [inline lib/gui/os/window-progress/services/window-progress.js]: Lol

File: lib/gui/pages/main/controllers/flash.js [hunk 12]
```diff
@@ -111,9 +111,12 @@ module.exports = function(
       return 'Flash!';
     }
 
-    if (flashState.percentage === 0 && !flashState.speed) {
+    const PERCENTAGE_MINIMUM = 0;
```
Comment (lurch) [inline lib/gui/pages/main/controllers/flash.js]: Does the same constant in multiple files imply that it out to be abstracted out?

File: lib/gui/pages/main/controllers/flash.js [hunk 13]
```diff
@@ -111,9 +111,12 @@ module.exports = function(
...
+    const PERCENTAGE_MAXIMUM = 100;
+
+    if (flashState.percentage === PERCENTAGE_MINIMUM && !flashState.speed) {
       return 'Starting...';
-    } else if (flashState.percentage === 100) {
+    } else if (flashState.percentage === PERCENTAGE_MAXIMUM) {
       if (isChecking && SettingsModel.get('unmountOnSuccess')) {
         return 'Unmounting...';
       }
```
Comment (lurch) [inline lib/gui/pages/main/controllers/flash.js]: There's multiple separate `if ...` blocks here. Should they be combined into a single `if ... else if ... else if ... else ...` block?

File: lib/gui/pages/settings/settings.js [hunk 14]
```diff
@@ -25,7 +25,8 @@ const MODULE_NAME = 'Etcher.Pages.Settings';
 const SettingsPage = angular.module(MODULE_NAME, [
   require('angular-ui-router'),
   require('../../components/warning-modal/warning-modal'),
-  require('../../models/settings')
+  require('../../models/settings'),
+  require('../../modules/error')
```
Comment (lurch) [inline lib/gui/pages/settings/settings.js]: Hmmm? How did the code work without this `require('../../modules/error')` before?

File: lib/image-stream/archive.js [hunk 15]
```diff
@@ -170,11 +170,12 @@ exports.extractImage = (archive, hooks) => {
   return hooks.getEntries(archive).then((entries) => {
 
     const imageEntries = _.filter(entries, (entry) => {
-      const extension = path.extname(entry.name).slice(1);
+      const extension = _.join(_.tail(path.extname(entry.name)), '');
```
Comment (lurch) [inline lib/image-stream/archive.js]: See my earlier `extension` comments

File: lib/shared/drive-constraints.js [hunk 16]
```diff
@@ -20,6 +20,14 @@ const _ = require('lodash');
 const pathIsInside = require('path-is-inside');
 
 /**
+ * @summary The default size for things such as images and drives
+ * @constant
+ * @private
+ * @type {Number}
+ */
+const DEFAULT_SIZE = 0;
```
Comment (lurch) [inline lib/shared/drive-constraints.js]: Would `UNKNOWN_SIZE` be better here?

File: lib/child-writer/index.js [hunk 17]
```diff
@@ -144,9 +162,13 @@ exports.write = (image, drive, options) => {
         });
       }
 
-      if (code !== EXIT_CODES.SUCCESS && code !== EXIT_CODES.VALIDATION_ERROR) {
-        return emitError(new Error(`Child process exited with error code: ${code}`));
+      // We shouldn't emit the `done` event manually here
+      // since the writer process will take care of it.
+      if (code === EXIT_CODES.SUCCESS || code === EXIT_CODES.VALIDATION_ERROR) {
```
Comment (jviotti) [inline lib/child-writer/index.js]: There are various ways the CLI process can fail, which are grouped into two categories:

- The CLI finished its task (flash & validation)
- The CLI crashed in the middle

If the CLI finished its task, then there are two possible ways out:

- The flash completed successfully (`EXIT_CODES.SUCCESS`)
- The flash completed, but the validation failed (`EXIT_CODES.VALIDATION_ERROR`)

If the CLI completes successfully, then it will emit its own `done` event, therefore we should not emit ourselves here.

File: lib/child-writer/renderer-utils.js [hunk 18]
```diff
@@ -40,9 +40,12 @@ exports.getApplicationEntryPoint = () => {
     return path.join(process.resourcesPath, 'app.asar');
   }
 
+  const ENTRY_POINT_ARGV_INDEX = 1;
+  const RELATIVE_ENTRY_POINT = electron.remote.process.argv[ENTRY_POINT_ARGV_INDEX];
```
Comment (jviotti) [inline lib/child-writer/renderer-utils.js]: Sounds like a good convention! I'll fix.

File: lib/child-writer/writer-proxy.js [hunk 19]
```diff
@@ -37,8 +37,32 @@ const packageJSON = require('../../package.json');
...
+ */
+const OPTIONS_INDEX_START = 2;
+
+/**
+ * @summary The list of Etcher argument options
+ * @constant
+ * @private
+ * @type {String[]}
+ */
+const ETCHER_ARGUMENTS = process.argv.slice(OPTIONS_INDEX_START);
```
Comment (jviotti) [inline lib/child-writer/writer-proxy.js]: Fixed!

File: lib/child-writer/writer-proxy.js [hunk 20]
```diff
@@ -169,6 +193,18 @@ return isElevated().then((elevated) => {
         child.on('error', reject);
         child.on('close', resolve);
 
+        /**
+         * @summary Emit an object message to the IPC server
+         * @function
+         * @private
+         *
+         * @param {Buffer} data - message data
```
Comment (jviotti) [inline lib/child-writer/writer-proxy.js]: It should be JSON. I clarified it on the param.

File: lib/cli/etcher.js [hunk 21]
```diff
@@ -53,7 +56,7 @@ isElevated().then((elevated) => {
       // If `options.yes` is `false`, pass `undefined`,
```
Comment (jviotti) [inline lib/cli/etcher.js]: Good catch!

File: lib/cli/etcher.js [hunk 22]
```diff
@@ -120,7 +124,7 @@ isElevated().then((elevated) => {
       return robot.printError(error);
     }
 
-    errors.print(error);
+    return errors.print(error);
```
Comment (jviotti) [inline lib/cli/etcher.js]: Haha, fixed!

File: lib/cli/options.js [hunk 23]
```diff
@@ -25,6 +25,25 @@ const EXIT_CODES = require('../shared/exit-codes');
 const packageJSON = require('../../package.json');
 
 /**
+ * @summary The required number of CLI arguments
+ * @constant
+ * @private
+ * @type {Number}
+ */
+const REQUIRED_NUMBER_OF_ARGUMENTS = 1;
```
Comment (jviotti) [inline lib/cli/options.js]: Yeah, I like how that reads.

File: lib/cli/options.js [hunk 24]
```diff
@@ -25,6 +25,25 @@ const EXIT_CODES = require('../shared/exit-codes');
...
+
+/**
+ * @summary The first index that represents an actual option argument
+ * @constant
+ * @private
+ * @type {Number}
+ *
+ * @description
+ * The first arguments are usually the program executable itself, etc.
```
Comment (jviotti) [inline lib/cli/options.js]: Not really. Consider that `argv` looks like this:

```sh
node bin/cli.js path/to/image.js <options>
```

The first two ones are not options.

Maybe the name is confusing? My intention is that `OPTIONS_INDEX_START` equals the first index of `argv` that is a option to the program.

File: lib/cli/options.js [hunk 25]
```diff
@@ -25,6 +25,25 @@ const EXIT_CODES = require('../shared/exit-codes');
...
+ */
+const OPTIONS_INDEX_START = 2;
```
Comment (jviotti) [inline lib/cli/options.js]: See https://github.com/resin-io/etcher/pull/1149#discussion_r104035788. This is not related to the minimum number of arguments.

File: lib/cli/options.js [hunk 26]
```diff
@@ -64,20 +83,21 @@ module.exports = yargs
...
     }
 
-    process.exit(1);
+    process.exit(EXIT_CODES.GENERAL_ERROR);
   })
 
   // Assert that image exists
   .check((argv) => {
-    fs.accessSync(argv._[0]);
+    const IMAGE_PATH_ARGV_INDEX = 0;
```
Comment (jviotti) [inline lib/cli/options.js]: Done!

File: lib/gui/components/modal/services/modal.js [hunk 27]
```diff
@@ -60,17 +60,11 @@ module.exports = function($uibModal, $q) {
           .catch((error) => {
 
             // Bootstrap doesn't 'resolve' these but cancels the dialog;
```
Comment (jviotti) [inline lib/gui/components/modal/services/modal.js]: Fixed!

File: lib/gui/components/update-notifier/services/update-notifier.js [hunk 28]
```diff
@@ -60,7 +60,9 @@ module.exports = function($http, $q, ModalService, UPDATE_NOTIFIER_SLEEP_TIME, M
           // The error status equals -1 if the request couldn't
           // be made successfully, for example, because of a
           // timeout on an unstable network connection.
-          if (error.status === -1) {
+          const ERROR_CODE_UNSUCCESSFULL_REQUEST = -1;
```
Comment (jviotti) [inline lib/gui/components/update-notifier/services/update-notifier.js]: > IMHO this constant should either be moved to the top of this file, or to the constants.js that I saw used elsewhere. 

The reason I put it here is that this error code is only relevant in the context of the `etcherLatestVersion` function. I'm worried that moving it somewhere else (even if its the top of the file), will make it seem that it applies somewhere else.

> And the comment above should be changed to not refer directly to -1.

Good idea!

> And there's only one 'l' in 'unsuccessful' ;)

Ah, fixed!

File: lib/gui/etcher.js [hunk 29]
```diff
@@ -41,12 +41,13 @@ electron.app.on('ready', () => {
   // Prevent flash of white when starting the application
   // https://github.com/atom/electron/issues/2172
   mainWindow.webContents.on('did-finish-load', () => {
+    const FLASH_TIMEOUT_MS = 100;
```
Comment (jviotti) [inline lib/gui/etcher.js]: Makes sense. I've changed it to `WEBVIEW_LOAD_TIMEOUT_MS`, since the timeout is necessary because the Chromium WebView takes a bit to load completely.

File: lib/gui/models/drives.js [hunk 30]
```diff
@@ -64,19 +64,34 @@ Drives.service('DrivesModel', function() {
...
+   *
+   * The problem is that every time you call `myImmutableObject.toJS()`
+   * you will get a new object, whose reference is different from
+   * the one you previously got, even if the data is exactly the same.
+   *
+   * @param {Function} func - function that returns an ImmutableJS list
+   * @returns {Function} memoized function
+   *
+   * @example
+   * const memoizedFunction = memoizeImmutableListReference(() => {
```
Comment (jviotti) [inline lib/gui/models/drives.js]: Fixed!

File: lib/gui/models/store.js [hunk 31]
```diff
@@ -75,8 +75,22 @@ const ACTIONS = _.fromPairs(_.map([
   return [ message, message ];
 }));
 
-const storeReducer = (state, action) => {
-  state = state || DEFAULT_STATE;
+/**
+ * @summary The redux store reduced
```
Comment (jviotti) [inline lib/gui/models/store.js]: Fixed!

File: lib/gui/models/store.js [hunk 32]
```diff
@@ -75,8 +75,22 @@ const ACTIONS = _.fromPairs(_.map([
...
+ * @param {Object} currentState - application state
+ * @param {Object} action - dispatched action
+ * @returns {Object} new application state
+ *
+ * @example
+ * const newState = storeReducer(DEFAULT_STATE, {
+ *   type: ACTIONS.REMOVE_DRIVE
+ * });
+ */
+const storeReducer = (currentState, action) => {
```
Comment (jviotti) [inline lib/gui/models/store.js]: Oh, fixed! For some reason I though I could only use a default value for the last argument (?)

File: lib/gui/models/store.js [hunk 33]
```diff
@@ -75,8 +75,22 @@ const ACTIONS = _.fromPairs(_.map([
...
+  const state = currentState || DEFAULT_STATE;
```
Comment (jviotti) [inline lib/gui/models/store.js]: See https://github.com/resin-io/etcher/pull/1149#discussion_r104038893, I got confused about default arguments :)

File: lib/gui/models/store.js (hunk 1 above)
Comment (jviotti) [inline lib/gui/models/store.js]: These are property names declare for the default state at the top of the file. We *could* reduce the chances of typos by making constants but I wonder if its to much complexity for trying to mitigate the fact that JavaScript is not a strongly typed language. Maybe moving to TypeScript is the solution?

File: lib/gui/models/store.js [hunk 34]
```diff
@@ -91,7 +105,8 @@ const storeReducer = (state, action) => {
 
       const newState = state.set('availableDrives', Immutable.fromJS(action.data));
 
-      if (action.data.length === 1) {
+      const DRIVE_AUTOSELECT_REQUIRED_LENGTH = 1;
+      if (action.data.length === DRIVE_AUTOSELECT_REQUIRED_LENGTH) {
```
Comment (jviotti) [inline lib/gui/models/store.js]: Looks great, and much more obvious!

File: lib/gui/models/store.js (hunk 2 above)
Comment (jviotti) [inline lib/gui/models/store.js]: Fixed!

File: lib/gui/models/supported-formats.js (hunk 3 above)
Comment (jviotti) [inline lib/gui/models/supported-formats.js]: Sounds good. I used `_.replace`, but decided to not store another variable with the extension + the period, since its tricky to find a good name for each, so that its not confusing.

File: lib/gui/modules/analytics.js (hunk 4 above)
Comment (jviotti) [inline lib/gui/modules/analytics.js]: I rewrote it without using a one-liner. Its hopefully clearer now.

File: lib/gui/modules/drive-scanner.js (hunk 5 above)
Comment (jviotti) [inline lib/gui/modules/drive-scanner.js]: Done!

File: lib/gui/modules/drive-scanner.js (hunk 6 above)
Comment (jviotti) [inline lib/gui/modules/drive-scanner.js]: `.flatMap` is a utility from RxJS to make all the `drivelist` function calls originate from the same stream (otherwise RxJS creates a new one for each).

Regarding the other `.map()`, I divided it into two to hopefully make the transformations clearer. The first one updates the descriptions, and the other one applies the system filtering. Happy to bring it back to one if that was better.

File: lib/gui/os/dialog/services/dialog.js (hunk 10 above)
Comment (jviotti) [inline lib/gui/os/dialog/services/dialog.js]: Sounds like a good idea!

File: lib/gui/os/dialog/services/dialog.js (hunk 8 above)
Comment (jviotti) [inline lib/gui/os/dialog/services/dialog.js]: Haha, looks like I missed this completely. I created the `BUTTONS` array but was only using it to get the indexes :)

File: lib/gui/os/dialog/services/dialog.js (hunk 7 above)
Comment (jviotti) [inline lib/gui/os/dialog/services/dialog.js]: Yeah, I agree. The Electron dialog module makes heavy use of number indexes, which is terrible to deal with: https://electron.atom.io/docs/api/dialog/#dialogshowmessageboxbrowserwindow-options-callback

I guess we can remove the need for `BUTTON_CONFIRMATION_INDEX ` using your suggestion from https://github.com/resin-io/etcher/pull/1149#discussion_r104024047, and only keep the rejection one, which is at least nicer. What do you think?

File: lib/gui/os/dialog/services/dialog.js (hunk 9 above)
Comment (jviotti) [inline lib/gui/os/dialog/services/dialog.js]: Hm, but that might fail if both the confirmation and rejection label are the same, which would be terrible on its own, but maybe its better to be on the safe side?

File: lib/gui/os/window-progress/services/window-progress.js (hunk 11 above)
Comment (jviotti) [inline lib/gui/os/window-progress/services/window-progress.js]: Haha, the drawbacks of disallowing magic numbers :P

File: lib/gui/pages/main/controllers/flash.js (hunk 13 above)
Comment (jviotti) [inline lib/gui/pages/main/controllers/flash.js]: Done!

File: lib/gui/pages/main/controllers/flash.js (hunk 12 above)
Comment (jviotti) [inline lib/gui/pages/main/controllers/flash.js]: Yeah, but where? `lib/shared/percentage.js`? :P I guess we can start a file with constants in `lib/shared/constants.js`. What do you think?

File: lib/gui/pages/settings/settings.js (hunk 14 above)
Comment (jviotti) [inline lib/gui/pages/settings/settings.js]: Coincidence. Another module probably required the error module before this module started, so it all just worked. The Angular module system is awful though, and doesn't give any warning for these kind of things. React should make things better.

File: lib/image-stream/archive.js (hunk 15 above)
Comment (jviotti) [inline lib/image-stream/archive.js]: Fixed!

File: lib/shared/drive-constraints.js (hunk 16 above)
Comment (jviotti) [inline lib/shared/drive-constraints.js]: Yep, fixed!

File: lib/gui/components/update-notifier/update-notifier.js [hunk 35]
```diff
@@ -29,7 +29,18 @@ const UpdateNotifier = angular.module(MODULE_NAME, [
...
 
-UpdateNotifier.constant('UPDATE_NOTIFIER_SLEEP_TIME', 7 * 24 * 60 * 60 * 100);
+/**
+ * @summary The update notifier sleep time
+ * @constant
+ * @private
+ * @type {Number}
+ *
+ * @description
+ * Seven days: 7 * 24 * 60 * 60 * 100
```
Comment (jviotti) [inline lib/gui/components/update-notifier/update-notifier.js]: Crap, this went undetected for so long! :O Amazing catch!

File: lib/gui/components/update-notifier/update-notifier.js [hunk 36]
```diff
@@ -29,7 +29,18 @@ const UpdateNotifier = angular.module(MODULE_NAME, [
   require('../../os/open-external/open-external')
 ]);
 
-UpdateNotifier.constant('UPDATE_NOTIFIER_SLEEP_TIME', 7 * 24 * 60 * 60 * 100);
+/**
+ * @summary The update notifier sleep time
```
Comment (jviotti) [inline lib/gui/components/update-notifier/update-notifier.js]: Great catch! Done! I also realized we should move the days to milliseconds conversion to `units.js`.

File: lib/gui/components/update-notifier/services/update-notifier.js (hunk 28 above)
Comment (lurch) [inline lib/gui/components/update-notifier/services/update-notifier.js]: > this error code is only relevant in the context of the `etcherLatestVersion` function.

In that case maybe the same module that exports the `etcherLatestVersion` function could also export appropriate constants for the error codes? :)

File: lib/gui/models/store.js (hunk 32 above)
Comment (lurch) [inline lib/gui/models/store.js]: Hmmm, good point. You know javascript better than I do ;-)

File: lib/gui/components/update-notifier/services/update-notifier.js (hunk 28 above)
Comment (jviotti) [inline lib/gui/components/update-notifier/services/update-notifier.js]: Hm, I investigated this one in more detail and it turns out its a bit tricky. The originator of this `-1` error code when the request couldn't be made is `$http`. `etcherLatestVersion` accepts a  function that given a URL allows the client to determine how to perform an HTTP request to it (given we use the module in the browser as well).

Since `etcherLatestVersion` accepts a function to make HTTP requests, then hardcoding the `-1` knowledge there looks a bit wrong, since different HTTP request functions might use different error codes for different things.

That leaves us with a few options:

- Expose a set of error constants from `etcherLatestVersion` and update the `.catch` block of the `$http` promise to compare against `-1` and translate that to an error number `etcherLatestVersion` understands, but that means we still have to declare this `-1` constant here

- Make `etcherLatestVersion` ignore every error coming from the HTTP requester function. This means we will swallow errors that we should be seeing

- Leave it as it is now

What do you think?

**Comment by jviotti** [comment]
@lurch Do I have the green sign for this one? :)

File: lib/gui/components/modal/services/modal.js [hunk 37]
```diff
@@ -60,17 +60,11 @@ module.exports = function($uibModal, $q) {
...
-            // therefore call 'resolve' here applied to 'false'.
             if (error === 'escape key press' || error === 'backdrop click') {
-              resolve();
-
-            // For some annoying reason, UI Bootstrap Modal rejects
-            // the result reason if the user clicks on the backdrop
-            // (e.g: the area surrounding the modal).
-            } else if (error !== 'backdrop click') {
-              return reject(error);
+              return resolve();
```
Comment (Shou) [inline lib/gui/components/modal/services/modal.js]: Yeah, this works.

File: lib/gui/os/dropzone/directives/dropzone.js [hunk 38]
```diff
@@ -41,7 +41,7 @@ module.exports = ($timeout) => {
       osDropzone: '&'
     },
     link: (scope, element) => {
-      const domElement = element[0];
+      const domElement = _.first(element);
```
Comment (jhermsmeier) [inline lib/gui/os/dropzone/directives/dropzone.js]: Since `element` seems to be a node list, shouldn't it be `elements`?

File: .eslintrc.yml [hunk 39]
```diff
@@ -447,19 +529,33 @@ rules:
...
     - error
+  prefer-numeric-literals:
+    - error
+  prefer-rest-params:
+    - error
+  prefer-template:
+    - error
+  prefer-arrow-callback:
+    - error
+    - allowNamedFunctions: false
```
Comment (jhermsmeier) [inline .eslintrc.yml]: Does this mean `.method( function() {})` or `.on( 'event', function() {})` will be disallowed?

File: lib/gui/os/dropzone/directives/dropzone.js (hunk 38 above)
Comment (jviotti) [inline lib/gui/os/dropzone/directives/dropzone.js]: @jhermsmeier Its a jQuery lite object, in which you can access the native DOM element by accessing its first element (which I agree its weird)

File: .eslintrc.yml (hunk 39 above)
Comment (jviotti) [inline .eslintrc.yml]: It means that something like `var foo = function foo() { ... }` is not allowed.

Comment (jhermsmeier) [inline .eslintrc.yml]: Ah I see, `prefer-arrow-callback`'s semantics weren't quite clear to me

File: lib/gui/os/dropzone/directives/dropzone.js (hunk 38 above)
Comment (jhermsmeier) [inline lib/gui/os/dropzone/directives/dropzone.js]: Oh, right – in that case renaming it to `$element` could clarify it. What do you think?

Comment (jviotti) [inline lib/gui/os/dropzone/directives/dropzone.js]: Sounds good, fixed!

**Comment by jviotti** [comment]
The conflict hell just began :P

File: lib/cli/options.js (hunk 24 above)
Comment (lurch) [inline lib/cli/options.js]: > Maybe the name is confusing? My intention is that OPTIONS_INDEX_START equals the first index of argv that is a option to the program.

Yeah, I understood that, I just think the `@description` is confusing. Perhaps it'd simply be clearer if the `@description` here (and at https://github.com/resin-io/etcher/pull/1149/files#diff-1db67373c40ea3d2756fbc0768723000R54 ) was just removed. I don't think it adds anything.

File: lib/cli/options.js (hunk 25 above)
Comment (lurch) [inline lib/cli/options.js]: Oh? If `MINIMUM_NUMBER_OF_ARGUMENTS` is `1`, doesn't that mean that the program-usage looks like:
```
program_name compulsory_argument [optional_arguments, ...]
```
and so the fact that there's a single compulsory_ argument, means that the optional_arguments start at index 2, i.e. `MINIMUM_NUMBER_OF_ARGUMENTS + 1` ?

Whereas if `MINIMUM_NUMBER_OF_ARGUMENTS` is `2`, the program-usage would be:
```
program_name compulsory_argument1 compulsory_argument2 [optional_arguments, ...]
```
and `OPTIONS_INDEX_START` would then be `3`, which is still `MINIMUM_NUMBER_OF_ARGUMENTS + 1` ?

File: lib/gui/components/update-notifier/services/update-notifier.js (hunk 28 above)
Comment (lurch) [inline lib/gui/components/update-notifier/services/update-notifier.js]: Yeah, I guess you could leave it as it is now, but maybe add a comment explaining that the `-1` actually relates to the `$http` object, because currently that's kinda unclear. I guess  `$http` doesn't have any useful constants you could import here?
`ERROR_CODE_UNSUCCESSFUL_REQUEST` is kinda generic-sounding ;)  What about `ERROR_HTTP_REQUEST_UNSUCCESSFUL`?

(P.S. Sorry for coming back to this so late)

File: lib/gui/os/dialog/services/dialog.js [hunk 40]
```diff
@@ -149,19 +159,19 @@ module.exports = function($q, SupportedFormatsModel) {
    * OSDialogService.showError('Foo Bar', 'An error happened!');
    */
   this.showError = (error, description) => {
```
Comment (lurch) [inline lib/gui/os/dialog/services/dialog.js]: I guess this could use the same `this.showError = (error, description) => {` as https://github.com/resin-io/etcher/pull/1149#discussion_r104005634 ?

File: lib/gui/os/window-progress/services/window-progress.js [hunk 41]
```diff
@@ -47,11 +44,14 @@ module.exports = function() {
...
+
+    if (percentage > PERCENTAGE_MAXIMUM || percentage < PERCENTAGE_MINIMUM) {
```
Comment (lurch) [inline lib/gui/os/window-progress/services/window-progress.js]: Only a minor nit, but I think comparing the minimum value first reads more naturally? i.e.
```js
if (percentage < PERCENTAGE_MINIMUM || percentage > PERCENTAGE_MAXIMUM) {
```

File: lib/gui/models/store.js (hunk 1 above)
Comment (lurch) [inline lib/gui/models/store.js]: >  We could reduce the chances of typos by making constants but I wonder if it's too much complexity

\*shrug\*

> trying to mitigate the fact that JavaScript is not a strongly typed language

Even a strongly-typed language doesn't prevent making typos in strings-used-as-constants ;-P

> Maybe moving to TypeScript is the solution?

I'm still getting to grips with ES, I'm in no hurry to switch to TS ;-D

File: lib/gui/os/dialog/services/dialog.js (hunk 7 above)
Comment (lurch) [inline lib/gui/os/dialog/services/dialog.js]: Given that we've _only_ got two buttons here, could you just do:
```js
const BUTTON_CONFIRMATION_INDEX = 0;
const BUTTON_REJECTION_INDEX = 1;
```
rather than using `indexOf`? As you say below, this would currently go wrong if `confirmationLabel` and `rejectionLabel` were the same.

File: lib/gui/pages/main/controllers/flash.js (hunk 12 above)
Comment (lurch) [inline lib/gui/pages/main/controllers/flash.js]: `lib/shared/constants.js` sounds good :+1:

File: lib/gui/app.js [hunk 42]
```diff
@@ -164,7 +167,7 @@ app.run(($window, WarningModalService, ErrorService, FlashStateModel, OSDialogSe
     // Don't open any more popups
     popupExists = true;
 
-    return OSDialogService.showWarning({
+    OSDialogService.showWarning({
       confirmationLabel: 'Yes, quit',
       rejectionLabel: 'Cancel',
```
Comment (lurch) [inline lib/gui/app.js]: Now that `showWarning` applies default button labels, this line could theoretically be removed ;-)
(I'm not suggesting we do though, because it probably makes the code more unreadable)
